python/ray/workflow/workflow_state.py

```diff
@@ -0,0 +1,290 @@
+            tid = dag_visit_queue.popleft()
+            if tid in visited_nodes:
+                continue
+            visited_nodes.add(tid)
+            self.pending_input_set[tid] = set()
+            for in_task_id in self.in_dependency[tid]:
+                self.reference_set[in_task_id].add(tid)
+                # All upstream deps should already complete here,
+                # so we just check their checkpoints.
+                task_input = self.get_input(in_task_id)
+                if task_input is None:
+                    self.pending_input_set[tid].add(in_task_id)
+                    dag_visit_queue.append(in_task_id)
+            if tid in self.latest_continuation:
+                if self.pending_input_set[tid]:
+                    raise ValueError(
+                        "A task that already returns a continuation cannot be pending."
+                    )
+                # construct continuations, as they are not directly connected to
+                # the DAG dependency
+                self.construct_runtime_deps(self.latest_continuation[tid])
+            elif not self.pending_input_set[tid]:
+                self.append_frontier_to_run(tid)
+
+    def init_context(self, context: WorkflowStepContext) -> None:
+        """Initialize the context of all tasks."""
+        for task_id, option in self.task_options.items():
+            self.task_context.setdefault(
+                task_id,
+                dataclasses.replace(
+                    context, task_id=task_id, checkpoint=option.checkpoint
+                ),
+            )
+
+    def add_task_done_callback(self, task_id: StepID, callback: asyncio.Future) -> None:
+        """Add done task callback of a task."""
+        task_id = self.continuation_root.get(task_id, task_id)
+        output = self.get_input(task_id)
+        if output is not None:
+            callback.set_result(output)
+        elif task_id in self.done_tasks:
+            callback.set_exception(
+                ValueError(
+                    f"Task '{task_id}' is done but neither in memory or in storage "
+                    "could we find its output. It could because its in memory "
+                    "output has been garbage collected and the task did not"
+                    "checkpoint its output."
+                )
+            )
+        else:
+            # The task could be a continuation in the future, so we need to take care of
+            # this scenario  in the workflow executor.
+            self.done_task_callbacks[task_id].append(callback)
+
+    def is_untracked_task(self, task_id: StepID) -> bool:
+        """A task is untracked if
+
+        1. The task itself (options etc) is not recorded in the state.
+        2. The output of the task is not recorded in the state.
+        """
+        return task_id not in self.task_options and self.get_input(task_id) is None
+
+
+def initial_state_from_dag(dag: Workflow) -> WorkflowExecutionState:
```

Comment:
nit: I'm thinking about whether Workflow is still necessary. Or can we do better about the layering here? If we have a ray dag, do we still need to transform it to workflow object? Or can the workflow object a subclass of ray dag?

```
Ray DAG
Workflow
----------
Execution engine
```

I haven't checked it, but I'd like to make sure Workflow/Ray DAG is only about the frontend thing.

Reply:
The reason I'm thinking about this is that, we have similar logic in recovery.
So the representation of the DAG is different in different places. Could it improve the structure if we defined the protocol/IR well?

Reply:
I removed `Workflow` now.